eclair-core/src/main/scala/fr/acinq/eclair/db/pg/PgChannelsDb.scala

```diff
@@ -62,37 +64,68 @@ class PgChannelsDb(implicit ds: DataSource, lock: PgLock) extends ChannelsDb wit
           statement.executeUpdate("CREATE TABLE htlc_infos (channel_id TEXT NOT NULL, commitment_number BIGINT NOT NULL, payment_hash TEXT NOT NULL, cltv_expiry BIGINT NOT NULL, FOREIGN KEY(channel_id) REFERENCES local_channels(channel_id))")
+
+          statement.executeUpdate("CREATE INDEX local_channels_type_idx ON local_channels ((json->>'type'))")
+          statement.executeUpdate("CREATE INDEX local_channels_remote_node_id_idx ON local_channels ((json->'commitments'->'remoteParams'->>'nodeId'))")
           statement.executeUpdate("CREATE INDEX htlc_infos_idx ON htlc_infos(channel_id, commitment_number)")
         case Some(v@2) =>
           logger.warn(s"migrating db $DB_NAME, found version=$v current=$CURRENT_VERSION")
           migration23(statement)
           migration34(statement)
+          migration45(statement)
         case Some(v@3) =>
           logger.warn(s"migrating db $DB_NAME, found version=$v current=$CURRENT_VERSION")
           migration34(statement)
+          migration45(statement)
+        case Some(v@4) =>
+          logger.warn(s"migrating db $DB_NAME, found version=$v current=$CURRENT_VERSION")
+          migration45(statement)
         case Some(CURRENT_VERSION) => () // table is up-to-date, nothing to do
         case Some(unknownVersion) => throw new RuntimeException(s"Unknown version of DB $DB_NAME found, version=$unknownVersion")
       }
       setVersion(statement, DB_NAME, CURRENT_VERSION)
     }
   }
 
+  /** Sometimes we may want to do a full reset when we update the json format */
+  def resetJsonColumns(connection: Connection): Unit = {
+    migrateTable(connection, connection,
+      "local_channels",
+      "UPDATE local_channels SET json=?::JSONB WHERE channel_id=?",
+      (rs, statement) => {
+        val state = stateDataCodec.decode(BitVector(rs.getBytes("data"))).require.value
+        val json = serialization.writePretty(state)
+        statement.setString(1, json)
+        statement.setString(2, state.channelId.toHex)
+      }
+    )(logger)
+  }
+
   override def addOrUpdateChannel(state: HasCommitments): Unit = withMetrics("channels/add-or-update-channel", DbBackends.Postgres) {
     withLock { pg =>
       val data = stateDataCodec.encode(state).require.toByteArray
       using(pg.prepareStatement(
         """
-          | INSERT INTO local_channels (channel_id, data, is_closed)
-          | VALUES (?, ?, FALSE)
+          | INSERT INTO local_channels (channel_id, data, json, is_closed)
+          | VALUES (?, ?, ?::JSONB, FALSE)
           | ON CONFLICT (channel_id)
-          | DO UPDATE SET data = EXCLUDED.data ;
+          | DO UPDATE SET data = EXCLUDED.data, json = EXCLUDED.json ;
           | """.stripMargin)) { statement =>
         statement.setString(1, state.channelId.toHex)
         statement.setBytes(2, data)
+        statement.setString(3, serialization.writePretty(state))
```

Comment:
This will be bad for our performance as JSON is notoriously slow (and this happens at every channel update).
But I think at that point we're not chasing perf at all, the convenience and ease of administration it provides is worth it.

In the (long term) future, I think the "main" DB shouldn't have this column at all, only replicas used for administration should.

Reply:
I don't think the performance impact would be that big, because 1) we can parallelize the `scodec`/`json` serialization and 2) the serialization time is much smaller than the actual db write.

We could also make the json column optional.

> In the (long term) future, I think the "main" DB shouldn't have this column at all, only replicas used for administration should.

This would create complications, because then we would have to manually take care of updating the json column. It can't be delegated to db triggers since it requires custom json serializers.